app/reports/descriptive_shape.rb

```diff
@@ -0,0 +1,92 @@
+# frozen_string_literal: true
+
+# Generate a report on the "shape" of Cocina descriptive metadata:
+#
+# bin/rails r -e production "DescriptiveShape.report"
+#
+# This will output a CSV file of JSON paths that contain a string or number value:
+#
+# path,count
+# .purl,4452
+# .title[].value,4321
+# .form[].type,2163
+# .form[].value,1810
+# .form[].source.value,1358
+# .note[].value,1319
+# .event[].date[].type,825
+# .event[].date[].encoding.code,629
+# .event[].date[].qualifier,2
+# .event[].date[].structuredValue[].type,18
+# .event[].date[].structuredValue[].value,18
+# .event[].note[].type,219
+# .event[].note[].value,219
+# .event[].note[].source.value,203
+# .event[].location[].code,204
+# .event[].location[].source.code,204
+# .event[].date[].value,816
+# .event[].location[].value,196
+# .title[].note[].type,161
+# .title[].note[].value,155
+# .title[].status,178
+# .title[].structuredValue[].type,344
+# .title[].structuredValue[].value,344
+# ...
+#
+# The results will also be written as descriptive-shape.json
+#
+class DescriptiveShape
+  def self.report
+    new.report
+  end
+
+  def initialize
+    @shape = Hash.new(0)
+  end
+
+  def report
+    Dro.find_each do |obj|
+      trace(obj.description)
+    end
+
+    save
+    output
+  end
+
+  private
+
+  def trace(obj, path = '')
+    case obj
+    when Array
+      trace_array(obj, path)
+    when Hash
+      trace_hash(obj, path)
+    else
+      @shape[path] += 1
+    end
+  end
+
+  def trace_array(obj, path)
+    obj.each do |item|
+      trace(item, "#{path}[]")
+    end
+  end
+
+  def trace_hash(obj, path)
+    obj.each do |key, value|
+      trace(value, "#{path}.#{key}")
+    end
+  end
+
+  def save
+    File.write('descriptive-shape.json') do |fh|
+      fh.write(JSON.pretty_generate(@shape))
+    end
```

Comment:
I *think* you can do this without the block?

```suggestion
    File.write('descriptive-shape.json', JSON.pretty_generate(@shape))
```